src/csharp/Grpc.Core/Internal/BatchContextSafeHandle.cs

```diff
@@ -54,21 +57,26 @@ public IntPtr Handle
             }
         }
 
-        public BatchCompletionDelegate CompletionCallback { get; set; }
+        public void SetCompletionCallback(BatchCompletionDelegate callback, object state)
+        {
+            GrpcPreconditions.CheckState(completionCallbackData.Callback == null);
+            GrpcPreconditions.CheckNotNull(callback, nameof(callback));
+            completionCallbackData = new CompletionCallbackData(callback, state);
```

Comment:
The only thing odd is it seems `CompletionCallbackData` is kind of a replacement for the previous closure that was created before every call to `RegisterBatchCompletion`, and though the closure is not getting alloc'd anymore, `CompletionCallbackData` still is. It seems like there should be a way for the `BatchCompletionDelegate`'s to know about the `I*Callback`'s that they're wrapping without needing for them to be supplied with one by their caller.

rough idea: If the `BatchCompletionDelegate`'s were instance methods instead of static methods, then they should be able to call their own call's `I*Callback` methods without being supplied with a `state` parameter. Then, I believe those `I*Callback` methods would no longer need to be passed in to methods like e.g., `StartUnary`. I'm pretty sure that if the `CallSafeHandles` knew the `AsyncCall` that they held on to, then this could work.

Reply:
The difference is that `CompletionCallbackData` is a struct and hence no allocation - so this approach works (and I've verified in a microbenchmark that no allocation happens).

Comments on the idea you described:
- whenever the BatchCompletionDelegates point to instance methods (instead of static methods), they will somehow need to capture the value of `this` - so there necessarily needs to be an allocation. We also cannot store the BatchCompletionDelegates as instance fields (as opposed to static fields) because many of the objects (CallSafeHandle, AsyncCall, ...) are likely to be created in large quantities and wouldn't want to have that overhead when creating new calls.